DATA_FORMAT.md

```diff
@@ -0,0 +1,330 @@
+# Nucmorph-Colorizer Data Format
+
+Nucmorph-Colorizer can only load datasets that follow the defined data specification.
+
+The easiest way to get started is to modify one of our existing data processing scripts, like [`generate_data.py`](./timelapse-colorizer-data/generate_data.py)!
+
+## Terms
+
+Here are a few important terms:
+
+- **Track**: A segmented object that has been tracked across one or more frames in the time series.
+- **Dataset**: A dataset is a single time-series, and can have any number of tracked objects and features.
+- **Collection**: An arbitrary grouping of datasets.
+
+## 📁 Dataset
+
+A dataset consists of a group of files that describe the tracks, feature data, processed images, and additional metadata for a single time-series.
+
+The most important file is the **manifest**, which describes all the files in the dataset.
+
+```
+--manifest.json--
+{
+    "frames": [
+        <relative path to image frame 0>,
+        <relative path to image frame 1>,
+        ...
+    ],
+    "features": {
+        <feature name 1>: <relative path to feature JSON>,
+        <feature name 2>: <relative path to feature JSON>,
+        ...
+    },
+    "outliers": <relative path to outlier JSON>,
+    "tracks": <relative path to tracks JSON>,
+    "times": <relative path to times JSON>,
+    "centroids": <relative path to centroids JSON>,
+    "bounds": <relative path to bounds JSON>
+}
+
+*Note: all paths are relative to this JSON file
+```
+
+<details>
+<summary><b>[Show me an example!]</b></summary>
+
+---
+
+An example dataset directory could look like this:
+
+```
+📂 my_dataset/
+  - 📄 manifest.json
+  - 📄 outliers.json
+  - 📄 tracks.json
+  - 📄 times.json
+  - 📄 centroids.json
+  - 📄 bounds.json
+  - 📕 feature_0.json
+  - 📗 feature_1.json
+  - 📘 feature_2.json
+  - 📁 frames/
+    - 📷 frame_0.png
+    - 📷 frame_1.png
+    - 📷 frame_2.png
+    ...
+    - 📷 frame_245.png
+```
+
+The `manifest.json` file would look something like this:
+
+```
+--manifest.json--
+{
+    "frames": [
+        "frames/frame_0.png",
+        "frames/frame_1.png",
+        "frames/frame_2.png",
+        ...
+        "frames/frame_245.png",
+    ],
+    "features": {
+        "My Cool Feature": "feature_0.json",
+        "Another Cool Feature": "feature_1.json",
+        "The Coolest Feature": "feature_2.json",
+    },
+    "outliers": "outliers.json",
+    "tracks": "tracks.json",
+    "times": "times.json",
+    "centroids": "centroids.json",
+    "bounds": "bounds.json"
+}
+```
+
+---
+
+</details>
+
+### 1. Tracks
+
+Every segmented object in each time step has a **track ID**, an integer identifier that's unique across all time steps. To recognize a single track across multiple frames, these track IDs must be grouped together with a single **track number**.
```

Comment:
Every segmented object belongs to a track but also has an id that is unique across all segmented objects in the entire time series.  These object ids are required to be sequential starting from 0.  For example: frame 1 may have objects 0, 1, 2.  Frame 2 has objects 3,4,5.  The tracking algorithm has determined that objects 0 and 3 are the same object, and belong to track 149.   Objects 1 and 4 are track 163.  And so on.  The track ids are not required to be sequential.

Reply:
Typically our data sources are spreadsheets with one row per unique segmented object, in which case the most convenient and natural way to assign object ids is simply by row number.

Reply:
Got it, I'll replace all the instances of "track ID" with "object ID" to clarify. 

Are object ID's required to be sequential? I thought that was typical but not guaranteed.

Reply:
They have to be sequential because that's how we index into the feature arrays.  Otherwise we'd have to have some other kind of mapping.  It keeps things much simpler.  If users really needed to have more meaningful object id's we'd probably introduce a new json file that contained that list.  But for the purposes of our app, the object id is its row index.

Reply:
I see, I think I was mixing up object IDs with tracks here.
So across all tracks, object ID's must be sequential, but a track is not guaranteed to have sequential/adjacent object IDs?

Reply:
right, if a track is the set of object id's across all frames that track the same REAL object, then it may appear to be a random sequence of unique object ids.   And no object id can appear in more than 1 track.